From 676937ebc536889c1c64ce0068b3b8e2e3a0684a Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Thu, 15 Feb 2024 09:20:45 +0200
Subject: [PATCH] [FSW-402] fs1: add memory io functions

This patch adds memory read/write functions for cfg space
registers and fw memories.

Change-Id: I15ac61834d50819931c622ee6428d01c1c45b7f0
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 7c94dda..691442c 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -5,6 +5,7 @@
  */
 
 #include <boot_rom.h>
+#include <mem_io.h>
 #include <fs1_regs.h>
 #include <fs1_reg_map.h>
 #include <fs1_reg_mask.h>
diff --git a/include/mem_io.h b/include/mem_io.h
new file mode 100644
index 0000000..e0ef2f0
--- /dev/null
+++ b/include/mem_io.h
@@ -0,0 +1,74 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2023 HabanaLabs Ltd.
+ */
+
+#ifndef _MEM_IO_H_
+#define _MEM_IO_H_
+
+#include <stdint.h>
+#include <soc.h>
+
+#define HL_FLASH_BASE				(HL_GEN_MEM_BASE + HL_FLASH_OFFSET)
+#define HL_SBROM_BASE				(HL_FW_MEM_BASE + HL_SBROM_OFFSET)
+#define HL_FWSRAM_BASE				(HL_FW_MEM_BASE + HL_FWSRAM_OFFSET)
+#define HL_SCRATCHPAD_SRAM_BASE			(HL_FW_MEM_BASE + HL_SCRATCHPAD_SRAM_OFFSET)
+
+/* In ARC address manipulation BIT[22:24] are for die id */
+#define HL_ARC_ADDR_DIE_ID_SHIFT		25
+#define HL_ARC_ADDR_DIE_ID_MASK			(0x7 << HL_ARC_ADDR_DIE_ID_SHIFT)
+
+#define BIT(nr)					(1 << (nr))
+extern struct boot_rom_global_s boot_rom_global;
+
+static inline void base_writel(uint32_t val, uint32_t offset)
+{
+	*((volatile uint32_t *)(offset)) = val;
+}
+
+static inline uint32_t base_readl(uint32_t offset)
+{
+	uint32_t reg = *((volatile uint32_t *)(offset));
+
+	return reg;
+}
+
+static inline uint32_t regmap_readl(uint32_t offset)
+{
+	uint32_t reg = base_readl((offset & ~(HL_ARC_ADDR_DIE_ID_MASK)) |
+				(boot_rom_global.die_id << HL_ARC_ADDR_DIE_ID_SHIFT));
+
+	return reg;
+}
+
+static inline void regmap_writel(uint32_t val, uint32_t offset)
+{
+	base_writel(val, (offset & ~(HL_ARC_ADDR_DIE_ID_MASK)) |
+				(boot_rom_global.die_id << HL_ARC_ADDR_DIE_ID_SHIFT));
+}
+
+static inline uint32_t sp_sram_readl(uint32_t offset)
+{
+	uint32_t reg = base_readl(HL_SCRATCHPAD_SRAM_BASE + offset);
+
+	return reg;
+}
+
+static inline void sp_sram_writel(uint32_t val, uint32_t offset)
+{
+	base_writel(val, HL_SCRATCHPAD_SRAM_BASE + offset);
+}
+
+static inline uint8_t regmap_readb(uint32_t offset)
+{
+	uint32_t reg = *((volatile uint32_t *)(HL_CFG_BASE + offset));
+
+	return (uint8_t)(reg & 0xFF);
+}
+
+static inline void regmap_writeb(uint8_t val, uint32_t offset)
+{
+	*((volatile uint32_t *)(HL_CFG_BASE + offset)) = (uint32_t)val;
+}
+
+#endif /* _MEM_IO_H_ */
